sockeye/data_io.py

```diff
@@ -592,23 +568,54 @@ def get_num_shards(num_samples: int, samples_per_shard: int, min_num_shards: int
     return max(int(math.ceil(num_samples / samples_per_shard)), min_num_shards)
 
 
-def save_shard(shard_idx: int, data_loader: RawParallelDatasetLoader,
-               shard_sources: List[str], shard_targets: List[str],
-               shard_stats: 'DataStatistics', output_prefix: str, keep_tmp_shard_files: bool):
+def save_shard(shard_idx: int,
+               data_loader: RawParallelDatasetLoader,
+               shard_sources: List[str],
+               shard_targets: List[str],
+               source_vocabs: List[vocab.Vocab],
+               target_vocabs: List[vocab.Vocab],
+               length_ratio_mean: float,
+               length_ratio_std: float,
+               buckets: List[Tuple[int, int]],
+               output_prefix: str,
+               keep_tmp_shard_files: bool):
     """
-    Load shard source and target data files into NDArrays and save to disk.
+    Load raw shard source and target data files, map to integers using the corresponding vocabularies,  
+    convert data into NDArrays and save to disk.
     Optionally it can delete the source/target files.
 
     :param shard_idx: The index of the shard.
     :param data_loader: A loader for loading parallel data from sources and target.
     :param shard_sources: A list of source file names.
     :param shard_targets: A list of target file names.
-    :param shard_stats: The statistics for the sources/target data.
+    :param source_vocabs: Source vocabulary (and optional source factor vocabularies).
+    :param target_vocabs: Target vocabulary (and optional target factor vocabularies).
+    :param length_ratio_mean: Mean length ratio.
+    :param length_ratio_std: Standard deviation of length ratios.
+    :param buckets: Bucket list.
     :param output_prefix: The prefix of the output file name.
     :param keep_tmp_shard_files: Keep the sources/target files when it is True otherwise delete them.
+    :return: Shard statistics.
     """
-    sources_sentences = [SequenceReader(s) for s in shard_sources]
-    targets_sentences = [SequenceReader(s) for s in shard_targets]
+
+    # Compute shard statistics and bucketing
+    shard_stat_accumulator = DataStatisticsAccumulator(buckets, source_vocabs[0], target_vocabs[0],
+                                                       length_ratio_mean, length_ratio_std)
+
+    # Shards contain the raw sentences. Need to map to integers using the vocabs and add BOS/EOS
+    sources_sentences = [SequenceReader(source, vocab, add_eos=True) for source, vocab in zip(shard_sources, source_vocabs)]
```

Comment:
Could you call `create_sequence_readers`?

```suggestion
    sources_sentences, target_sentences = create_sequence_readers(shard_sources, shard_targets, source_vocabs, target_vocabs)
```